rpcs3/rpcs3qt/rsx_debugger.cpp

```diff
@@ -1202,3 +1215,103 @@ void rsx_debugger::GetBuffers() const
 
 	m_buffer_tex->showImage(QImage(m_buffer_tex->cache.data(), width, height, QImage::Format_RGB32));
 }
+
+void rsx_debugger::GetVertexProgram() const
+{
+	const auto render = rsx::get_current_renderer();
+	if (!render || !render->is_initialized || !render->local_mem_size || !render->is_paused())
+	{
+		return;
+	}
+
+	RSXVertexProgram vp;
+	vp.data.reserve(512 * 4);
+
+	const u32 vp_entry = rsx::method_registers.transform_program_start();
+	const auto vp_metadata = program_hash_util::vertex_program_utils::analyse_vertex_program
+	(
+		rsx::method_registers.transform_program.data(),  // Input raw block
+		vp_entry,                                        // Address of entry point
+		vp                                               // [out] Program object
+	);
+
+	const u32 ucode_len = static_cast<u32>(vp.data.size() * sizeof(u32));
+	std::vector<u32> vp_blob = {
+		7003u,             // Type
+		6u,                // Revision
+		56u + ucode_len,   // Total size
+		0,                 // paramCount
+		0,                 // paramArray
+		32u,               // Program header offset
+		ucode_len,         // Ucode length
+		56u,               // Ucode start
+
+		::size32(vp.data) / 4, // Instruction count
+		0,                     // Slot
+		16u,                   // Registers used
+		rsx::method_registers.vertex_attrib_input_mask(),
+		rsx::method_registers.vertex_attrib_output_mask(),
+		rsx::method_registers.clip_planes_mask()
+	};
+
+	vp_blob.resize(vp_blob.size() + vp.data.size());
+	std::copy(vp.data.begin(), vp.data.end(), vp_blob.begin() + 14);
+
+	std::span<u32> vp_binary(vp_blob);
+	CgBinaryDisasm vp_disasm(vp_binary);
+	vp_disasm.BuildShaderBody(false);
+
+	m_transform_disasm->clear();
+	m_transform_disasm->setText(QString::fromStdString(vp_disasm.GetArbShader()));
+}
+
+void rsx_debugger::GetFragmentProgram() const
+{
+	const auto render = rsx::get_current_renderer();
+	if (!render || !render->is_initialized || !render->local_mem_size || !render->is_paused())
+	{
+		return;
+	}
+
+	const auto [program_offset, program_location] = rsx::method_registers.shader_program_address();
+	auto data_ptr = vm::base(rsx::get_address(program_offset, program_location));
```

Comment:
This must be using vm::get_super_ptr to avoid access violstions on the main thread. And additional check of vm::check_addr for the same reason.